### PR TITLE
Group14

### PR DESCRIPTION
**Pull request by lunncurt**
The ```docs/group14.md``` gives an accurate state of the module including current relevant classes to review (or at least main contributors to current game state).

Main changes this submission were:
```
  - demos/InteractiveWorld/Group14_main.cpp
  - source/Worlds/Hub/InteractiveWorld.hpp
  - source/Agents/AI/FetchAgent.hpp
  - source/Agents/Classic/ResourceManagementAgent.hpp/cpp
  - source/Worlds/Hub/Building.hpp
  - source/Worlds/Hub/ResourceProducer.hpp
```
  After that, these are the main supporting files:
```
  - source/Worlds/Hub/ResourceSpawn.hpp
  - source/Worlds/Hub/ResourceBank.hpp
  - source/Worlds/Hub/TownHall.hpp
  - source/Interfaces/TrashInterface.hpp
  - source/Worlds/Hub/InteractiveWorldSaveManager.hpp
```

### CODE REVIEW

**Comment by karensuzue** [comment]
**Reviewers for Group 14:** Please select all commits starting from **[0542cc9](https://github.com/CSE498/Spring2026-CompanyA/pull/91/commits/0542cc9087c8d1d712b2658150fbcdb68ddf7f83) and earlier**

**Review by AB123456kkkkk — COMMENTED** [review]
Testing and Error Checking
TestFarmingAgent — Overall there are unit tests for every single public member function (including constructors). The test cases are also well organized.  Most of the private portions from the header and cpp files are included in the code. One edge case that is tested is null pointers.
TestFetchAgent — Unit tests for all public member functions (including constructors). The test cases are also well organized.  Most of the private portions from the header and cpp files are included in the code. One edge case that is tested out of bounds. No asserts in this file and TestFarmingAgent.
ResourceBank — Unit tests for all public member functions (including constructors). The test cases are also well organized.  Most of the private portions from the header and cpp files are included in the code. Edge case include boundary conditions.
TestResourceManagementAgent — This test case is different than the others one, mainly due to more checks in the file.
Overall I liked the fact that there is good unit testing in the code. All of the test cases are well organized and trigger cases that exercise the private portions. All of the edge cases have been tested. Error conditions are checked well. Good error handling. 


Implementation
All of the code is under the cse498 namespace, which is good practice. The files do not have any namespace pollution and no use of using namespace std.  No unnecessary code duplication is there, appropriate modularity is maintained. Great use of memory management — used shared pointers and unique pointers throughout all of the files and no use of raw pointers. That is one of things that I liked about the code is the use of memory management. Const correctness is excellent throughout the files. Member functions are marked as const and parameters are also marked as const. For modern C++ constructs — there are uses of lambdas (2). Used templates. Multiple uses of std::optional, no uses of constexpr. Mixed both range-based for loops and standard uses. I was able to run the code perfectly — no errors or warnings during compilation. The code avoids undefined behavior. Used nodiscard whenever applicable. There are uses of algorithms.

File: demos/InteractiveWorld/Group14_main.cpp [hunk 1]
```diff
@@ -5,14 +5,17 @@
  **/
```
Comment (murra381) [inline demos/InteractiveWorld/Group14_main.cpp:6]: Code Design: Clear enough to understand. The configureFetcher lambda improves readability. I thought that was good.

Advanced C++: Uses std::make_shared, std::make_unique, and a lambda well.

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 2]
```diff
@@ -5,6 +5,7 @@
  **/
```
Comment (murra381) [inline source/Worlds/Hub/InteractiveWorld.hpp:6]: Code Design: Interface is understandable, I thought AddBuilding, AddResourceSpawn, AddResourceBank, and AddTownHall were easy to read and simple. TryInteractAdjacent is readable. I liked the small change in PrintInventory as well.

Advanced C++: Good use of std::array for interaction priority and std::shared_ptr/std::make_shared for inventory.

File: source/Agents/AI/FetchAgent.hpp [hunk 3]
```diff
@@ -0,0 +1,331 @@
+/**
+ * This file is for the Fall 2026 CSE 498 section 2 Capstone project.
+ * @brief Agent that shuttles between an origin point and a deposit point.
+ */
```
Comment (murra381) [inline source/Agents/AI/FetchAgent.hpp:4]: Code Design: Strong interface overall. My main concern is the raw pointer ownership in m_origin and m_depositPoint; the code appears to assume those endpoint agents outlive the FetchAgent. I’d like to see that lifetime contract documented more clearly or represented more safely if possible.

Advanced C++: Good use of std::function, lambdas, [[nodiscard]], and structured default behavior. The BFS implementation is solid. No major notes here. Good job.

File: source/Agents/Classic/ResourceManagementAgent.cpp [hunk 4]
```diff
@@ -0,0 +1,548 @@
+/**
+ * @file ResourceManagementAgent.cpp
+ * @brief Interactive world agent for upgrading buildings and selling resources.
```
Comment (murra381) [inline source/Agents/Classic/ResourceManagementAgent.cpp:3]: Code Design: The public API is functional and easy to follow, with clear operations for upgrading buildings, selling resources, hiring lanes, and querying lane state. The upgrade, sell, and hire logic is separated into methods that can be tested directly, which is a strong design choice. Good job.

Advanced C++: Good use of std::optional, std::string* optional output messages, and aliases like GoldAmount. The parsing helpers are a nice small modular choice too. Good job with that.

File: source/Worlds/Hub/Building.hpp [hunk 5]
```diff
@@ -7,6 +7,7 @@
 #pragma once
```
Comment (murra381) [inline source/Worlds/Hub/Building.hpp:8]: Code Design: This was good. Had upgradeable building plus local resource storage. Good use of private validation helper.  m_rateModifier = 0.25 could probably be a named static constexpr.

Advanced C++: Uses std::expected, std::optional, constexpr, and value semantics for BuildingUpgrade. This was good.

File: source/Worlds/Hub/ResourceProducer.hpp [hunk 6]
```diff
@@ -10,7 +10,7 @@
 #include <memory>
 
 #include "Building.hpp"
```
Comment (murra381) [inline source/Worlds/Hub/ResourceProducer.hpp:12]: Code Design: Small nitpick but m_outputType seems to be stored and never used after creation. It also stores raw pointers to Building and ResourceSpawn. Could maybe use references or non-owning pointers, but that is another nitpick.

Advanced C++: Uses std::chrono well. Could use constexpr defaults or stronger types for production rates, but current use is reasonable.

File: source/Worlds/Hub/ResourceSpawn.hpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
+/**
+ * This file is for the Fall 2026 CSE 498 section 2 Capstone project.
```
Comment (murra381) [inline source/Worlds/Hub/ResourceSpawn.hpp:2]: Code Design: The interface is simple and easy to understand, but I would add validation for negative values. AddResource() should reject negative quantities, and SetMaxCollectionQuantity() should reject negative limits. Otherwise Collect() can return a negative amount or accidentally increase the stored resource count. What exactly is stopping me from inputting -5 to AddResource()?

Advanced C++: Could benefit from static constexpr int DEFAULT_MAX_COLLECTION_QUANTITY = 10.

File: source/Worlds/Hub/ResourceSpawn.hpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+ * @brief Represents a spawn point for resources that the Agents will grab from
+ * @note Status: PROPOSAL
```
Comment (murra381) [inline source/Worlds/Hub/ResourceSpawn.hpp:4]: ResourceBank.hpp
Code Design: I thought the code design for this class was clean.

Advanced C++: Uses a type alias from InteractiveWorldInventory, [[nodiscard]], and composition cleanly. Good design.

File: source/Worlds/Hub/TownHall.hpp [hunk 9]
```diff
@@ -0,0 +1,47 @@
+/**
+ * This file is part of the Fall 2026, CSE 498, section 2, course project.
```
Comment (murra381) [inline source/Worlds/Hub/TownHall.hpp:2]: Code Design: This one needs some work. The class comment says @class Building, which is incorrect. The constructor that leaves m_inventory null makes DepositResource() unsafe. If someone creates a TownHall using the constructor without an inventory, then m_inventory is nullptr. Calling DepositResource() after that would dereference a null pointer, which can crash or cause undefined behavior.

Advanced C++: Uses std::shared_ptr and std::map.

File: source/Worlds/Hub/InteractiveWorldSaveManager.hpp [hunk 10]
```diff
@@ -4,6 +4,7 @@
 #include <string>
 
 #include "../../../third-party/json/json.hpp"
```
Comment (murra381) [inline source/Worlds/Hub/InteractiveWorldSaveManager.hpp:6]: Code Design: Useful single-purpose class. The save and load responsibilities are easy to understand, and the code is organized clearly around inventory, buildings, and resource-manager state. I thought this class was good overall.

Advanced C++: Uses nlohmann::json cleanly and makes good use of range-based loops. The implementation is straightforward and readable.

**Review by murra381 — COMMENTED** [review]
Sorry this is a little late.

File: source/Worlds/Hub/InteractiveWorld.hpp
Comment (finfroc3) [inline source/Worlds/Hub/InteractiveWorld.hpp]: TODO: API??

File: source/Interfaces/TrashInterface.hpp
Comment (finfroc3) [inline source/Interfaces/TrashInterface.hpp]: TODO: API??
is this one needed even

File: tests/Agents/TestResourceManagementAgent.cpp
Comment (finfroc3) [inline tests/Agents/TestResourceManagementAgent.cpp]: TODO: test??

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 11]
```diff
@@ -23,12 +26,13 @@ namespace cse498 {
 class InteractiveWorld : public WorldBase {
 protected:
     // World Inventory
-    InteractiveWorldInventory m_inventory;
+    std::shared_ptr<InteractiveWorldInventory> m_inventory = std::make_shared<InteractiveWorldInventory>();
```
Comment (marksandychez) [inline source/Worlds/Hub/InteractiveWorld.hpp:29]: Good use of a shared_ptr when an object will be referenced by multiple classes

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 12]
```diff
@@ -60,59 +91,83 @@ class InteractiveWorld : public WorldBase {
...
     void PrintInventory() {
         std::ostringstream output;
-        output << m_inventory.GetAmount(ItemType::Wood) << ' ' << ItemTypeToString(ItemType::Wood) << " | "
-               << m_inventory.GetAmount(ItemType::Stone) << ' ' << ItemTypeToString(ItemType::Stone) << " | "
-               << m_inventory.GetAmount(ItemType::Metal) << ' ' << ItemTypeToString(ItemType::Metal);
+        output << m_inventory->GetAmount(ItemType::Wood) << ' ' << ItemTypeToString(ItemType::Wood) << " | "
+               << m_inventory->GetAmount(ItemType::Stone) << ' ' << ItemTypeToString(ItemType::Stone) << " | "
+               << m_inventory->GetAmount(ItemType::Metal) << ' ' << ItemTypeToString(ItemType::Metal);
 
         std::cout << output.str() << std::endl;
```
Comment (marksandychez) [inline source/Worlds/Hub/InteractiveWorld.hpp:98]: Could potentially upgrade to use print for more modern C++ usage, though this has zero effect on actual code execution.

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 13]
```diff
@@ -60,59 +91,83 @@ class InteractiveWorld : public WorldBase {
...
 
+
 public:
     /**
      * Constructor
      */
     InteractiveWorld() {
         floor_id = main_grid.AddCellType("floor", "Floor that agents can walk on.", ' ');
         wall_id = main_grid.AddCellType("wall", "Impenetrable wall.", '#');
+        building_id = main_grid.AddCellType("building", "An impassable building.", 'B');
```
Comment (marksandychez) [inline source/Worlds/Hub/InteractiveWorld.hpp:126]: "Floor" and "Wall" could cause issues with the GUI team, though this can be further discussed with them. Image assets require unique "name" implementation to be drawn.

File: source/Agents/AI/FetchAgent.hpp [hunk 14]
```diff
@@ -0,0 +1,306 @@
...
+            if (it->second == from) {
+                break;
+            }
+            step = it->second;
+        }
+
+        if (step == from.Up()) {
+            return GetActionID("up");
+        }
+        if (step == from.Down()) {
```
Comment (marksandychez) [inline source/Agents/AI/FetchAgent.hpp:293]: Could this be shorted to an if-elif structure?

File: source/Agents/Classic/FarmingAgent.cpp [hunk 15]
```diff
@@ -1,8 +1,242 @@
...
+        {
+            // building level can scale how much stock farmer replenishes
+            amount += mAssignedBuilding->GetCurrentLevel();
+        }
+
+        RestockLimitedOffer(mRestockItemName, amount);
+
+        // Reset work timer and start heading home after work
+        mTicksSinceWork = 0;
+        mState = FarmerState::ReturningHome;
```
Comment (marksandychez) [inline source/Agents/Classic/FarmingAgent.cpp:176]: Good use of an enumeration

File: source/Agents/Classic/FarmingAgent.cpp [hunk 16]
```diff
@@ -1,8 +1,242 @@
...
+    }
+
+    std::size_t FarmingAgent::SelectAction(const WorldGrid& grid)
+    {
+        ++mTicksSinceWork;
+
+        if (mAssignedBuilding == nullptr)
```
Comment (marksandychez) [inline source/Agents/Classic/FarmingAgent.cpp:183]: I believe the Google Style Guide usually prefers implicit pointer checks like:
`if (ptr) { }`
But this does not effect code

File: source/Worlds/Hub/ResourceSpawn.hpp [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+     */
+    size_t SelectAction(const WorldGrid&) override { return 0; }
+
+    /**
+     * Add resources to the pool
+     * @param quantity quantity to add
+     */
+    void AddResource(const int& quantity) {
+        // Check for integer overflow
+        if (quantity > std::numeric_limits<int>::max() - m_quantity)
```
Comment (marksandychez) [inline source/Worlds/Hub/ResourceSpawn.hpp:52]: Great edge case checking, though mathematically I'd be impressed if this was hit. Still great standard practice

File: source/Worlds/Hub/Building.hpp [hunk 18]
```diff
@@ -99,6 +103,18 @@ class Building {
      * @return current level as int
      */
     int GetCurrentLevel() const { return m_level; }
+    /**
+     * Set the current building level
+     * @param level level to set building to
+     * @return bool if setting level was successful
+     */
+    [[maybe_unused]] bool SetCurrentLevel(int level) {
```
Comment (marksandychez) [inline source/Worlds/Hub/Building.hpp:111]: This feels like something that maybe shouldn't have it's return value ignored, and could be changed to [[nodiscard]]. If there is a reason, I might just be missing it

**Review by marksandychez — COMMENTED** [review]
# Advanced C++
A large amount of these classes revolve around integer calculation and proper resource transitions, which don't leave much room for overly many advanced C++ features. Despite this, I would consider the chrono clock to resource generation an example of a well implemented advanced feature. There are also a variety of shared_ptrs in cases where Buildings will need to refer to a common inventory object. The code also makes great use attributes to indicate the relevance of the return values of the functions. There might be potential for use of a lambda when checking the results of up, down, left, and right returns, but is not truly needed for the simple boolean checks.

# Implementation Quality

- **No Warnings During Compilation**
With the most recent changes compiled and merged onto group 17s branch, there were no error stemming from any code.

- **Proper Use of Namespaces**
Files properly indicate cse498, and avoids pitfalls and bad practices like using namespace std

- **Adheres to the C++ standard**
Member variables are initialized and undefined behavior is avoided. Also makes use of expected.

- **Appropriate Modularity**
Classes clearly serve a purpose, and there is not a case of "those who know, do" issues within member variables

- **Modern C++**
As mentioned above, expected is used and range based for loops are used when possible

- **Pointer Handling**
Makes use of good pointer memory management by using shared and unique pointers when relevant. Null pointer checks are also used to avoid de-referencing invalid objects.


Great job!

File: tests/InteractiveWorld/Building.cpp
Comment (finfroc3) [inline tests/InteractiveWorld/Building.cpp:1]: All tests pass and look good overall, though SetCurrentLevel isn't tested (it is marked maybe_unused, though).  GetRateModifier, and SetRateModifier aren't included either, and though the setter is used in the test for ResourceProducer, the getter is never properly tested.

File: source/Interfaces/TrashInterface.hpp [hunk 19]
```diff
@@ -83,25 +86,34 @@ class TrashInterface : public InterfaceBase {
...
+                    InteractiveWorldSaveManager save_manager;
+                    if (save_manager.Save(*interactive_world, "interactive_world_save.json")) {
+                        std::cout << "\nGame saved to interactive_world_save.json\n";
+                    } else {
+                        std::cout << "\nFailed to save game.\n";
+                    }
+                }
+            }
+
                 exit(0); // Quit!
```
Comment (hoangtu0701) [inline source/Interfaces/TrashInterface.hpp:117]: Calling exit(0) here skips all destructors and RAII cleanup. If any other module needs to do something on shutdown it just won't happen. Setting a flag to break out of the game loop would be a lot safer.

File: source/Interfaces/TrashInterface.hpp [hunk 20]
```diff
@@ -83,25 +86,34 @@ class TrashInterface : public InterfaceBase {
...
-            case 'D':
                 action_id = GetActionID("right");
                 break;
-            case 'q':
-            case 'Q':
+            case 'e':
+                action_id = GetActionID("interact");
+                break;
+            case 'q': {
+                const auto* interactive_world = dynamic_cast<const InteractiveWorld*>(&world);
```
Comment (hoangtu0701) [inline source/Interfaces/TrashInterface.hpp:106]: This ties TrashInterface to InteractiveWorld specifically, which breaks the generic interface/world separation. If save-on-quit is needed, it'd be better to push that logic into the world's shutdown path or use a callback so the interface stays world-agnostic

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 21]
```diff
@@ -40,10 +44,37 @@ class InteractiveWorld : public WorldBase {
...
+
+    bool TryInteractAdjacent(const WorldPosition& position, const AgentBase* initiator) {
+        const std::array<WorldPosition, 4> priorities = {
+                position.Right(),
+                position.Up(),
+                position.Down(),
+                position.Left(),
+        };
+
+        for (const WorldPosition& target: priorities) {
```
Comment (hoangtu0701) [inline source/Worlds/Hub/InteractiveWorld.hpp:58]: This scans every agent for each adjacent tile, so it's O(agents * 4) per interaction. Works fine with a small agent count but a position-to-agent map would make this way cheaper if things scale up

File: source/Worlds/Hub/InteractiveWorld.hpp [hunk 22]
```diff
@@ -177,27 +217,55 @@ class InteractiveWorld : public WorldBase {
...
 
     /**
-     * Add NPC to the world
-     * @param npc NPC to add
+     * Add Town hall to the world
+     * @param th town hall
+     * @param position where to add spawn
      */
-    void AddNPC(std::shared_ptr<NPC> npc) { m_npcs.push_back(npc); }
+    void AddTownHall(TownHall& th, WorldPosition position) {
```
Comment (hoangtu0701) [inline source/Worlds/Hub/InteractiveWorld.hpp:260]: These four methods all do the same thing - set location and mark the tile. Could consolidate into a single method to cut the duplication.

File: source/Worlds/Hub/ResourceProducer.hpp [hunk 23]
```diff
@@ -72,14 +75,14 @@ class ResourceProducer {
...
+            m_resourceSpawn->AddResource(whole);
             m_accumulator -= whole;
         }
     }
 
 private:
-    InteractiveWorldInventory& m_inventory; // World Inventory
-    std::shared_ptr<Building> m_building; // Building modifying the output rate
+    ResourceSpawn* m_resourceSpawn;
+    Building* m_building = nullptr; // Building modifying the output rate
```
Comment (hoangtu0701) [inline source/Worlds/Hub/ResourceProducer.hpp:85]: Raw pointer here but the class doesn't own the building. Since the demo always passes references, storing a reference instead would make the lifetime relationship clearer.

File: source/Agents/AI/FetchAgent.hpp [hunk 24]
```diff
@@ -0,0 +1,331 @@
...
+        const long long dy =
+            static_cast<long long>(a.CellY()) - static_cast<long long>(b.CellY());
+        return std::llabs(dx) + std::llabs(dy) == 1;
+    }
+
+    /// @brief BFS through walkable tiles to find the first step that moves
+    ///        `from` closer to a tile adjacent to `goal`. The goal tile itself
+    ///        is treated as reachable even though it is non-walkable, so the
+    ///        expansion terminates on the tile adjacent to the endpoint.
+    [[nodiscard]] size_t NextStepToward(const WorldGrid& grid, WorldPosition from,
```
Comment (hoangtu0701) [inline source/Agents/AI/FetchAgent.hpp:262]: Full BFS every tick is a lot of work for agents that shuttle the same route over and over. Caching the path until the goal or grid changes would be a nice optimization.

**Review by hoangtu0701 — COMMENTED** [review]
Interoperability & API

Group14_main.cpp does a nice job showing the whole resource pipeline working together - spawns, producers, fetch agents, buildings, town hall, resource management menu, all in one runnable demo. The group14.md doc is also really solid, gives a clear breakdown of what each class does and how to test things manually.

A few things I noticed on the interoperability side:

1. TrashInterface calls exit(0) directly when the user hits 'q'. That skips all destructors and RAII cleanup, so any other module that needs to do something on shutdown (analytics logging, save confirmation, etc.) just never gets a chance to run. Would be cleaner to set a flag that breaks out of the game loop instead.

2. TrashInterface downcasts to InteractiveWorld inside SelectAction to handle saving. That means the interface now has a hard dependency on a specific world implementation, which kind of defeats the purpose of having a generic interface/world separation. Maybe move the save logic into the world's shutdown path or use a callback so the interface doesn't need to know what world it's in.

3. A few files still have `@note Status: PROPOSAL` but the classes are clearly functional and being used in the demo. Might want to update those.



Code Design

The class hierarchy makes sense overall. ResourceBank as a base for Building is a clean choice, the FetchAgent endpoint callback pattern is flexible, and the ResourceManagementAgent menu system is well structured. Some things that caught my eye:

1. InteractiveWorld::TryInteractAdjacent loops through every agent in the world for each of the 4 adjacent tiles. Works fine now but that's O(agents * 4) per interaction, which could get rough if the agent count grows. Even a simple position-to-agent map would bring that down to constant time per tile.

2. AddBuilding, AddResourceSpawn, AddResourceBank, and AddTownHall all do essentially the same thing - set the entity's location and mark the tile as building_id. Could probably consolidate those into one method to cut down on the repetition.

3. FetchAgent::NextStepToward runs a full BFS every single tick. For agents that are just shuttling back and forth on the same route, caching the path and only recalculating when the goal or grid actually changes would save a bunch of unnecessary work.

4. ResourceProducer holds a raw pointer to Building, but the demo always constructs producers with references. Storing a reference instead (or at least documenting the lifetime expectation) would make the non-ownership clearer and reduce the chance of a dangling pointer down the road.

5. FarmingAgent.hpp and FarmingAgent.cpp are Group 2's files. Having another group's source in your PR diff makes it a bit harder to tell what's actually Group 14's work vs. what's coming from elsewhere.

File: tests/InteractiveWorld/ResourceBank.cpp
Comment (finfroc3) [inline tests/InteractiveWorld/ResourceBank.cpp:1]: All tests pass and everything in ResourceBank is covered.

File: tests/InteractiveWorld/ResourceProducer.cpp
Comment (finfroc3) [inline tests/InteractiveWorld/ResourceProducer.cpp:1]: All tests pass and look good, could be a good idea to make tests for ResourceSpawn as well.

File: tests/Agents/TestFetchAgent.cpp
Comment (finfroc3) [inline tests/Agents/TestFetchAgent.cpp:1]: All tests pass and test the overall functionality and intended behavior well. If there's time, I suggest including additional tests for the pathfinding (things like obstacles or having there be no valid path to origin/deposit).

File: tests/Agents/TestResourceManagementAgent.cpp
Comment (finfroc3) [inline tests/Agents/TestResourceManagementAgent.cpp:1]: All tests pass, what is tested looks good but it is overall missing coverage for parts of ResourceManagementAgent. There exists nothing for hireable lanes, locked buildings, the functions for handling interactions, and nothing for minor functions like ClearManagedBuildings or some of the setters and getters.

File: demos/InteractiveWorld/Group14_main.cpp
Comment (finfroc3) [inline demos/InteractiveWorld/Group14_main.cpp:1]: Very good example driver! It runs just fine on my system. The agents, world, buildings and inventory are integrated very well, and the doc that came with was really detailed and well written. Thanks to that, I was quickly able to follow and understand the setup of everything. I do think that more comments are needed, though.

File: source/Agents/AI/FetchAgent.hpp
Comment (finfroc3) [inline source/Agents/AI/FetchAgent.hpp:1]: API is good, but more thorough documentation would be appreciated.

File: source/Agents/Classic/ResourceManagementAgent.hpp
Comment (finfroc3) [inline source/Agents/Classic/ResourceManagementAgent.hpp:1]: The API is mostly well designed and self explanatory, though complex. It greatly lacks in any documentation, though. The use of raw pointers is a little sketchy as well.

File: source/Worlds/Hub/InteractiveWorld.hpp
Comment (finfroc3) [inline source/Worlds/Hub/InteractiveWorld.hpp:1]: Outside of issues touched on in the other review comments, the API is well made and easy to understand.

File: source/Worlds/Hub/ResourceProducer.hpp
Comment (finfroc3) [inline source/Worlds/Hub/ResourceProducer.hpp:1]: Very simple, cleanly made, and well documented.

File: source/Worlds/Hub/Building.hpp
Comment (finfroc3) [inline source/Worlds/Hub/Building.hpp:1]: Very well documented, good job.
